python/lsst/ctrl/mpexec/cmdLineFwk.py

```diff
@@ -329,22 +285,108 @@ def makePipeline(self, taskFactory, args):
+            Pipeline, can be empty or ``None`` if graph is read from pickle
+            file.
+        taskFactory : `~lsst.pipe.base.TaskFactory`
+            Task factory.
+        args : `argparse.Namespace`
+            Parsed command line
+
+        Returns
+        -------
+        graph : `~lsst.pipe.base.QuantumGraph`
+        """
+        if args.qgraph:
+
+            with open(args.qgraph, 'rb') as pickleFile:
+                qgraph = pickle.load(pickleFile)
+
+            # pipeline cann not be provided in this case
+            if pipeline:
+                raise ValueError("Pipeline must not be given when quantum graph is read from file.")
+
+        else:
+
+            if not pipeline:
+                raise ValueError("Pipeline must be given for quantum graph construction.")
+
+            # build collection names
+            inputs = args.input.copy()
+            defaultInputs = inputs.pop("", None)
+            outputs = args.output.copy()
+            defaultOutputs = outputs.pop("", None)
+
+            # Make butler instance. From this Butler we only need Registry
+            # instance. Input/output collections are handled by pre-flight
+            # and we don't want to be constrained here by Butler's restrictions
+            # on collection names.
+            collection = defaultInputs[0] if defaultInputs else None
+            butler = Butler(config=args.butler_config, collection=collection)
+
+            # if default input collections are not given on command line then
+            # use one from Butler (has to be configured in butler config)
+            if not defaultInputs:
+                defaultInputs = [butler.collection]
+            coll = DatasetOriginInfoDef(defaultInputs=defaultInputs,
+                                        defaultOutput=defaultOutputs,
+                                        inputOverrides=inputs,
+                                        outputOverrides=outputs)
+
+            # make execution plan (a.k.a. DAG) for pipeline
+            graphBuilder = GraphBuilder(taskFactory, butler.registry, args.skip_existing)
+            qgraph = graphBuilder.makeGraph(pipeline, coll, args.data_query)
+
+        if args.save_qgraph:
+            with open(args.save_qgraph, "wb") as pickleFile:
+                pickle.dump(qgraph, pickleFile)
+
+        if args.qgraph_dot:
+            graph2dot(qgraph, args.qgraph_dot)
+
+        return qgraph
+
+    def runPipeline(self, graph, taskFactory, args):
         """Execute complete QuantumGraph.
 
         Parameters
         ----------
         graph : `QuantumGraph`
             Execution graph.
-        butler : `Butler`
-            data butler instance
+        taskFactory : `~lsst.pipe.base.TaskFactory`
+            Task factory.
         args : `argparse.Namespace`
             Parsed command line
         """
+        # If output collections are given then use them to override
+        # butler-configured ones.
```

Comment:
There can only be one output collection, right?

Reply:
I don't know 🤔 There is one default output collection, but I think we also support per-dataset type collections too. I'll change the comment to make it more explicit.

Reply:
I suppose I shouldn't have phrased this as a question - I don't think Butler can handle multiple output collections in the same run.